### PR TITLE
{Doc} Track 2 Migration Guidance

### PR DESCRIPTION
**Pull request by qwordy**
**Description**

{Doc} Track 2 Migration Guidance

Azure CLI is built upon Azure Python SDK. Recently Azure Python SDK announced next generation product. It is named Track 2 SDK. The old version of SDK is named Track 1. It claims that it has great advantages than Track 1 SDK. It is not compatible with Track 1 SDK. Azure CLI developers need to spend considerable time and do some work to migrate from Track 1 SDK to Track 2 SDK.

This documentation summarizes typical issues and solutions when adopting Track 2 SDK in Azure CLI.

You are welcome to contribute to this documentation if you have experience of Track 2 SDK.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
@jsntcy for awareness.

File: doc/track_2_migration_guidance.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Track 2 Migration Guidance
+
+Azure CLI is built upon Azure Python SDK. Recently Azure Python SDK announced next generation product. It is named Track 2 SDK. The old version of SDK is named Track 1. It claims that it has great advantages than Track 1 SDK. It is not compatible with Track 1 SDK. Azure CLI developers need to spend considerable time and do some work to migrate from Track 1 SDK to Track 2 SDK.
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:3]: suggest to put link of Track2 SDK if user interested at benefits from Track2 vs Track1.

File: doc/track_2_migration_guidance.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+This documentation summarizes typical issues and solutions when adopting Track 2 SDK in Azure CLI.
+
+**Long running operation function name change**
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:7]: Could you pls add a TOC of issue list for easy navigation?

File: doc/track_2_migration_guidance.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+Long running operations have changed their function names in Track 2 SDK. A `begin_` prefix is added. E.g. `create_or_update` becomes `begin_create_or_update`. `delete` becomes `begin_delete`. It is a naming convention in Track 2 SDK to indicate that an operation is a long running operation.
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:9]: how about add a PR link with code change to show detail for each item?

File: doc/track_2_migration_guidance.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+**Property name change**
+
+Some of property names change in Track 2 SDK.
+
+Examples:
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:15]: sample as above, sample PR is very helpful

File: doc/track_2_migration_guidance.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+hyperVgeneration -> hyperVGeneration
+disk_mbps_read_only -> disk_m_bps_read_only
+disk_mbps_read_write -> disk_m_bps_read_write
+virtual_machine_extension_type -> type_properties_type
+type1 -> type_properties_type
+instance_ids -> vm_instance_i_ds
+diskMbpsReadWrite -> diskMBpsReadWrite
+```
+
+Some changes are unnecessary, even wrong in English.
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:27]: not necessary

File: doc/track_2_migration_guidance.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+Examples:
+
+```
+VirtualMachineIdentityUserAssignedIdentitiesValue -> UserAssignedIdentitiesValue
+```
+
+**Error type change**
+
+Examples:
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:41]: wrong line?

File: doc/track_2_migration_guidance.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+Pay attention to error type.
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:43]: Error type changed in Track2 SDK.

File: doc/track_2_migration_guidance.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+Examples:
+
+In VMSS `begin_update_instances`, a new type `VirtualMachineScaleSetVMInstanceRequiredIDs` is added.
+
+In DiskAccess `begin_create_or_update`, location and tags are moved to a nested structure `DiskAccess`, `disk_access = DiskAccess(location=location, tags=tags)`
+
+**Subscription**
+
+There are various ways to obtain subscription ID. Some ways that work in Track 1 SDK do not work in Track 2 SDK.
```
Comment (yungezz) [inline doc/track_2_migration_guidance.md:67]: could you pls be specify about `Some ways`?

File: doc/track_2_migration_guidance.md (hunk 5 above)
Comment (yonzhan) [reply doc/track_2_migration_guidance.md:27]: Above changes should be fixed right? Do we have issue to track this?

**Review by yonzhan — COMMENTED** [review]
@jsntcy we can combine our findings into this Guidance.

File: doc/track_2_migration_guidance.md (hunk 5 above)
Comment (qwordy) [reply doc/track_2_migration_guidance.md:27]: > 
> 
> not necessary

why
https://dictionary.cambridge.org/us/dictionary/english/unnecessary

Comment (qwordy) [reply doc/track_2_migration_guidance.md:27]: > 
> 
> Above changes should be fixed right? Do we have issue to track this?

Yes. Yes. I opened one. https://github.com/Azure/autorest.python/issues/850

File: doc/track_2_migration_guidance.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+4. Error type change
+5. No enum type
+6. Class hierarchy change
+7. Obtaining Subscription
+8. Multi-API support
+9. Fixing mocked object
+10. Modifying patch_models.py to include missing packages
+11. Missing x-ms-authorization-auxiliary
+
+**Long running operation function name change**
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:25]: We can use `##...` to mark these lines as headings and link to them with [URL fragment](https://developer.mozilla.org/en-US/docs/Web/API/URL#Properties):

```
https://developer.mozilla.org/en-US/docs/Web/API/URL#Properties
                                                     ^^^^^^^^^^                                                       
```

File: doc/track_2_migration_guidance.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+hyperVgeneration -> hyperVGeneration
+disk_mbps_read_only -> disk_m_bps_read_only
+disk_mbps_read_write -> disk_m_bps_read_write
+virtual_machine_extension_type -> type_properties_type
+type1 -> type_properties_type
+instance_ids -> vm_instance_i_ds
+diskMbpsReadWrite -> diskMBpsReadWrite
+```
+
+Some changes are unnecessary, even wrong in English. I opened an [issue](https://github.com/Azure/autorest.python/issues/850) to track this problem.
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:45]: It's better to use the GitHub convention to represent an issue. `https://github.com/Azure/autorest.python/issues/850` will be converted to https://github.com/Azure/autorest.python/issues/850.

```suggestion
Some changes are unnecessary, even wrong in English. I opened [Azure/autorest.python#850](https://github.com/Azure/autorest.python/issues/850) to track this problem.
```

File: doc/track_2_migration_guidance.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+**Fixing mocked object**
+
+The problem I met is property name change. It is hard to find the line of code that causes the error. Please update mocked object construction code and make sure it is up-to-date.
+
+**Modifying patch_models.py to include missing packages**
+
+It is only used in CI jobs. It patches some code to SDK. This file should be deprecated. It was written long time ago. But for now, just modify this file and add missing packages.
+
+**Missing x-ms-authorization-auxiliary**
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:107]: "Missing external tenant credential support" would be more accurate.

File: doc/track_2_migration_guidance.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+2. Property name change
+3. Class name change
+4. Error type change
+5. No enum type
+6. Class hierarchy change
+7. Obtaining Subscription
+8. Multi-API support
+9. Fixing mocked object
+10. Modifying patch_models.py to include missing packages
+11. Missing x-ms-authorization-auxiliary
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:23]: We can use hyperlink to link to each section.

File: doc/track_2_migration_guidance.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+The class hierarchy may change in Track 2 SDK. Some properties are not flattened. They are wrapped in classes. 
+
+Examples:
+
+In VMSS `begin_update_instances`, a new type `VirtualMachineScaleSetVMInstanceRequiredIDs` is added.
+
+In DiskAccess `begin_create_or_update`, location and tags are moved to a nested structure `DiskAccess`, `disk_access = DiskAccess(location=location, tags=tags)`
+
+### Obtaining Subscription
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:81]: Better to use base form of verbs.

Comment (qwordy) [reply doc/track_2_migration_guidance.md]: http://www.davidnicholson.it/language/on-ing-forms-in-titles.html
I prefer gerund.

**Review by jiasli — APPROVED** [review]

File: doc/track_2_migration_guidance.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
+# Track 2 Migration Guidance
+
+Azure CLI is built upon Azure Python SDK. Recently Azure Python SDK announced next generation product. It is named Track 2 SDK. The old version of SDK is named Track 1. It claims that it has great advantages than Track 1 SDK. It is not compatible with Track 1 SDK. Azure CLI developers need to spend considerable time and do some work to migrate from Track 1 SDK to Track 2 SDK. Let's see an example of Track 2 SDK. [azure-mgmt-compute 17.0.0b1](https://pypi.org/project/azure-mgmt-compute/17.0.0b1/) introduces important breaking changes and important new features like unified authentication and asynchronous programming.
+
+This document summarizes typical issues and solutions when adopting Track 2 SDK in Azure CLI.
+
+Example PRs:
+1. [Compute PR](https://github.com/Azure/azure-cli/pull/15750)
+2. [Network PR](https://github.com/Azure/azure-cli/pull/16245)
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:9]:
```suggestion
1. [Compute PR #15750](https://github.com/Azure/azure-cli/pull/15750)
2. [Network PR #16245](https://github.com/Azure/azure-cli/pull/16245)
```

File: doc/track_2_migration_guidance.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+It is only used in CI jobs. It patches some code to SDK. This file should be deprecated. It was written long time ago. But for now, just modify this file and add missing packages.
+
+### Missing external tenant authentication support
+
+External tenant authentication support is missing in Azure CLI Core and Azure Core package. In request header, we use x-ms-authorization-auxiliary to pass auxiliary authorization token. Compute module is the first customer to have this requirement in Track 2 SDK. In azure/core/pipeline/policies/_authentication.py, there is a class BearerTokenCredentialPolicy. It simplifies bearer token authorization header configuration. However, auxiliary bearer token authorization is not supported in Azure Core policies [1] yet. The current solution is getting tokens manually, setting headers in client constructor or in operation call time manually. They said they will support this policy in the future. Azure CLI Core also needs an upgrade to provide better interfaces for Track 2 SDK users.
+
+You are welcome to contribute to this document if you have experience of Track 2 SDK.
+
+**References**
```
Comment (jiasli) [inline doc/track_2_migration_guidance.md:113]: This can be a heading.

File: doc/track_2_migration_guidance.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (qwordy) [inline doc/track_2_migration_guidance.md:113]:
```suggestion
### References
```

**Pull request merged**